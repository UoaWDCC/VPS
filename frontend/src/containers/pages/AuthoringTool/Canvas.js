import React from "react";

import styles from "../../../styling/Canvas.module.scss";

export default function Canvas() {
  return (
    <>
      <div className={styles.canvasContainer} />
    </>
  );
}
