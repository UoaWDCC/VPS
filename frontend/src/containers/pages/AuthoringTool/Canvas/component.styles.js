import { makeStyles } from "@material-ui/core";

export default function useStyles(component) {
  return makeStyles({
    testButton: {
      position: "absolute",
      height: `${component.height}`,
    },
  });
}
