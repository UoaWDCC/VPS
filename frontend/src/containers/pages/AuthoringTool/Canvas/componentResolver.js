import React from "react";
import ButtonComponent from "../Components/ButtonComponent/ButtonComponent";
import TextComponent from "../Components/TextComponent";
import ImageComponent from "../Components/ImageComponent/ImageComponent";

export default function componentResolver(component, id, selectElement) {
  switch (component.type) {
    // ADD NEW COMPONENT TYPES HERE
    case "BUTTON":
      return (
        <ButtonComponent
          id={id}
          selectElement={selectElement}
          component={component}
        />
      );
    case "TEXT":
      return (
        <TextComponent
          id={id}
          selectElement={selectElement}
          component={component}
        />
      );
    case "IMAGE":
      return (
        <ImageComponent
          id={id}
          selectElement={selectElement}
          component={component}
        />
      );
    default:
      break;
  }

  return <div />;
}
