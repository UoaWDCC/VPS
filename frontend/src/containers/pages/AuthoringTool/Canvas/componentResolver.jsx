/* eslint-disable no-param-reassign */
import ButtonComponent from "../Components/ButtonComponent";
import FirebaseAudioComponent from "../Components/FirebaseAudioComponent";
import FirebaseImageComponent from "../Components/FirebaseImageComponent";
import ImageComponent from "../Components/ImageComponent";
import SpeechTextComponent from "../Components/SpeechTextComponent";
import TextComponent from "../Components/TextComponent";

/**
 * This function returns the appropriate React component for a scene component object when editing
 *
 * @example
 * {currentScene?.components?.map((component, index) =>
 *   componentResolver(component, index, selectElement)
 * )}
 */
export default function componentResolver(component, index, onClick) {
  switch (component.type) {
    // ADD NEW COMPONENT TYPES HERE
    case "BUTTON":
      return (
        <ButtonComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "SPEECHTEXT":
      return (
        <SpeechTextComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "TEXT":
      return (
        <TextComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "IMAGE":
      return (
        <ImageComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "FIREBASEIMAGE":
      return (
        <FirebaseImageComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "FIREBASEAUDIO":
      return (
        <FirebaseAudioComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    default:
      break;
  }

  return <div />;
}
