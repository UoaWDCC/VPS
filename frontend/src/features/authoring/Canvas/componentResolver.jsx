import { MenuItem, MenuList, Paper } from "@material-ui/core";
import RightContextMenu from "../../../components/ContextMenu/RightContextMenu";
import ButtonComponent from "../components/ButtonComponent";
import FirebaseAudioComponent from "../components/FirebaseAudioComponent";
import FirebaseImageComponent from "../components/FirebaseImageComponent";
import ImageComponent from "../components/ImageComponent";
import SpeechTextComponent from "../components/SpeechTextComponent";
import TextComponent from "../components/TextComponent";
import SceneContext from "../../../context/SceneContext";
import { handleSendToBack } from "../CanvasSideBar/ComponentProperties/ZAxis";
import { useContext } from "react";
import { handle } from "../../../components/ContextMenu/portal";




const SceneMenu = ({ updateComponentProperty, currentScene, component }) => {
  return (
    <Paper>
      <MenuList>
        <MenuItem>Duplicate</MenuItem>
        <MenuItem onClick={() => handle(handleSendToBack, { currentScene, component, updateComponentProperty })}>Send To Back</MenuItem>
        <MenuItem>Delete</MenuItem>
      </MenuList>
    </Paper>
  );
};

/**
 * This function returns the appropriate React component for a scene component object when editing
 *
 * @example
 * {currentScene?.components?.map((component, index) =>
 *   componentResolver(component, index, selectElement)
 * )}
 */
export default function componentResolver(component, index, onClick) {
  const { currentScene, updateComponentProperty } = useContext(SceneContext);

  switch (component.type) {
    // ADD NEW COMPONENT TYPES HERE
    case "BUTTON":
      return (
        <ButtonComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "RESET_BUTTON":
      return (
        <ButtonComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "SPEECHTEXT":
      return (
        <SpeechTextComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "TEXT":
      return (
        <RightContextMenu menu={SceneMenu({ updateComponentProperty, currentScene, component })}>
          <TextComponent
            key={component.id}
            id={index}
            onClick={onClick}
            component={component}
          />
        </RightContextMenu>
      );
    case "IMAGE":
      return (
        <ImageComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "FIREBASEIMAGE":
      return (
        <FirebaseImageComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "FIREBASEAUDIO":
      return (
        <FirebaseAudioComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    default:
      break;
  }

  return <div />;
}
