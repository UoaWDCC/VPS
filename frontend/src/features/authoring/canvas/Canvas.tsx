import React, { useRef } from "react";
import CanvasContext from "./CanvasContext";
import Overlay from "./Overlay";
import type { Component } from "../types";
import TextBox from "../elements/TextBox";
import Speech from "../elements/Speech";
import Ellipse from "../elements/Ellipse";
import Box from "../elements/Box";
import Image from "../elements/Image";
import Line from "../elements/Line";
import useVisualScene from "../stores/visual";
import { handleMouseDownGlobal, handleMouseMoveGlobal, handleMouseUpGlobal } from "../handlers/pointer/pointer";
import { handleContextGlobal } from "../handlers/pointer/context";

const componentMap: Record<string, React.FC<any>> = {
  textbox: (props) => <TextBox {...props} editable={true} />,
  speech: Speech,
  ellipse: Ellipse,
  box: Box,
  image: Image,
  line: Line,
};

function resolve(component: Component) {
  const Fc = componentMap[component.type];
  if (Fc) return <Fc key={component.id} {...component} />;
  return null;
}

function Canvas() {
  const scene = useVisualScene(state => state.components);

  const canvasRef = useRef<SVGSVGElement | null>(null);

  if (!scene) return <></>;

  function toSVGSpace(cx: number, cy: number) {
    const boundingRect = canvasRef.current?.children[0];
    if (!boundingRect) return { x: 0, y: 0 };
    const { top, left, width, height } = boundingRect.getBoundingClientRect();
    const x = ((cx - left) / width) * 1920;
    const y = ((cy - top) / height) * 1080;
    return { x, y };
  }

  function handleMouseMove(e: React.MouseEvent) {
    handleMouseMoveGlobal(e, toSVGSpace(e.clientX, e.clientY));
  }

  function handleMouseUp() {
    handleMouseUpGlobal();
  }

  function handleMouseDown(e: React.MouseEvent) {
    handleMouseDownGlobal(e, toSVGSpace(e.clientX, e.clientY));
  }

  function handleContextMenu(e: React.MouseEvent) {
    handleContextGlobal(e, toSVGSpace(e.clientX, e.clientY));
  }

  const components = Object.values(scene).sort((a, b) => a.zIndex - b.zIndex).map(resolve);

  return (
    <CanvasContext.Provider value={{ toSVGSpace, canvasRef }} >
      <div
        className="flex-grow relative"
        onMouseMove={handleMouseMove}
        onMouseUp={handleMouseUp}
        onMouseDown={handleMouseDown}
        onContextMenu={handleContextMenu}
      >
        <Overlay />
        <svg
          id="main"
          className="w-full h-full"
          viewBox={`-50 -50 ${1920 + 50 * 2} ${1080 + 50 * 2}`}
          ref={canvasRef}
        >
          <rect x="0" y="0" width="1920" height="1080" fill="white" />
          {components}
        </svg>
      </div>
    </CanvasContext.Provider>
  );
}

export default Canvas;
