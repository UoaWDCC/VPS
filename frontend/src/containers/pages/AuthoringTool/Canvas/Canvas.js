/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable no-param-reassign */
import React, { useContext } from "react";

import Moveable from "react-moveable";
import AuthoringToolContext from "../../../../context/AuthoringToolContext";
import SceneContext from "../../../../context/SceneContext";
import styles from "../../../../styling/Canvas.module.scss";
import componentResolver from "./componentResolver";

export default function Canvas() {
  const { currentScene } = useContext(SceneContext);
  const {
    select,
    scalable,
    selectElement,
    clearElement,
    bounds,
    setBounds,
    shiftPressed,
    setShiftPressed,
    deleteElement,
  } = useContext(AuthoringToolContext);

  const keyDown = ({ key }) => {
    console.log(key);
    if (key === "Shift") {
      setShiftPressed(true);
    } else if (key === "Backspace") {
      deleteElement();
    }
  };

  const keyUp = ({ key }) => {
    if (key === "Shift") {
      setShiftPressed(false);
    }
  };

  document.addEventListener("keydown", keyDown);
  document.addEventListener("keyup", keyUp);

  return (
    <>
      <Moveable
        target={document.getElementById(select)}
        draggable
        throttleDrag={0}
        resizable={!scalable}
        scalable={scalable}
        keepRatio={shiftPressed}
        snappable
        bounds={bounds}
        onDragStart={() => {
          const canvas = document
            .getElementById("canvas")
            .getBoundingClientRect();
          setBounds({
            left: canvas.left,
            top: canvas.top,
            bottom: canvas.bottom,
            right: canvas.right,
          });
        }}
        onDrag={({ target, transform }) => {
          target.style.transform = transform;
        }}
        onDragEnd={({ target }) => {
          const canvas = document
            .getElementById("canvas")
            .getBoundingClientRect();
          const transfromMatrix = window
            .getComputedStyle(target)
            .transform.match(/(-?[0-9\\.]+)/g);
          if (transfromMatrix != null) {
            console.log(`x: ${(transfromMatrix[4] * 100) / canvas.width}`);
            console.log(`y: ${(transfromMatrix[5] * 100) / canvas.height}`);
          }
        }}
        onScale={({ target, scale, drag }) => {
          target.style.transform =
            `translate(${drag.beforeTranslate[0]}px, ${drag.beforeTranslate[1]}px)` +
            `scale(${scale[0]}, ${scale[1]})`;
        }}
        onScaleEnd={({ target }) => {
          const transfromMatrix = window
            .getComputedStyle(target)
            .transform.match(/(-?[0-9\\.]+)/g);
          if (transfromMatrix != null) {
            console.log(`x-scale: ${transfromMatrix[0]}`);
            console.log(`y-scale: ${transfromMatrix[3]}`);
          }
        }}
        onResize={({ target, width, height, drag }) => {
          target.style.width = `${width}px`;
          target.style.height = `${height}px`;
          target.style.transform = `translate(${drag.beforeTranslate[0]}px, ${drag.beforeTranslate[1]}px)`;
        }}
        onResizeEnd={({ target }) => {
          // cover case where no resizing occurs (onResize is not called)
          if (target.style.width.slice(-2) !== "px") {
            return;
          }
          const absWidth = Number(target.style.width.slice(0, -2));
          const absHeight = Number(target.style.height.slice(0, -2));
          const canvas = document
            .getElementById("canvas")
            .getBoundingClientRect();
          const relWidth = `${(absWidth / canvas.width) * 100}%`;
          const relHeight = `${(absHeight / canvas.height) * 100}%`;
          console.log(`width: ${relWidth}`);
          console.log(`height: ${relHeight}`);
          target.style.width = relWidth;
          target.style.height = relHeight;
        }}
      />

      <div className={styles.canvasContainer}>
        <div id="canvas" className={styles.canvas} onClick={clearElement}>
          {currentScene?.components?.map((component, index) =>
            componentResolver(component, index, selectElement)
          )}
        </div>
      </div>
    </>
  );
}
