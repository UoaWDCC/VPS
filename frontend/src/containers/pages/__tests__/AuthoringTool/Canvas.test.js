import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import Canvas from "../../AuthoringTool/Canvas";

test("Scenario Selection page snapshot test", () => {
  const component = renderer.create(
    <BrowserRouter>
      <Canvas />
    </BrowserRouter>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});