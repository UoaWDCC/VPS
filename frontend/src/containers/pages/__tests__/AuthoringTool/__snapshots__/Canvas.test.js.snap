// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scenario Selection page snapshot test 1`] = `
<div
  className="canvasContainer"
>
  <div
    className="canvas"
  />
</div>
`;
